### PR TITLE
Add multi-tenant routing and endpoints

### PR DESCRIPTION
**Pull request by dsbrgg**
This PR introduces a middleware for django called `DatabaseRouter` to be able to distinguish requests to their respective `domainId`.
A `/migrate` endpoint is also added to enable remote activation of migrations on a specified `domainId` sent through the request headers.

### CODE REVIEW

File: saleor/domain_utils.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import os
+import sys
+import json
+import dj_database_url
+
+from functools import wraps
+from urllib import request, parse
+
```
Comment (johnmaia) [inline saleor/domain_utils.py]: Remove new line.

File: saleor/domain_utils.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+from django.conf import settings
+from django.db import connections, transaction
+
```
Comment (johnmaia) [inline saleor/domain_utils.py]: Remove new line.

File: saleor/domain_utils.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    if request is not None:
+        return request.headers["X-Domain-ID"]
+    if "--domain" in opts:
+        index = opts.index("--domain") + 1
+        return opts[index]
+
+def update_migration_state(domain=None, state=None):
+    try:
+        payload = {'ecommerce_status': state}
```
Comment (johnmaia) [inline saleor/domain_utils.py]: `ecommerceStatus`

File: saleor/domain_utils.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        url = f"{saas_manager_host}/domains/{domain}"
+
```
Comment (johnmaia) [inline saleor/domain_utils.py]: Remove new line.

File: saleor/domain_utils.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            resp = response.read().decode('utf-8')
+            return json.loads(resp)
+    except Exception as e:
+        raise e
+
+def update_ecommerce_id(user_id=None, saleor_id=None):
+    try:
+        payload = {'ecommerceId': saleor_id}
+        url = f"{user_manager_host}/users/{user_id}"
+
```
Comment (johnmaia) [inline saleor/domain_utils.py]: Remove new line.

File: saleor/domain_utils.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        with request.urlopen(req) as response:
+            credentials = response.read().decode('utf-8')
+            return json.loads(credentials)
+    except Exception as e:
+        raise e
+
+def serialize_connection(credentials):
+    client = credentials.get("client")
+    connection = credentials.get("connection")
+
```
Comment (johnmaia) [inline saleor/domain_utils.py]: Remove new line.

File: saleor/domain_utils.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    password = connection.get("password")
+    database = connection.get("database")
+
+    return f"{client}://{user}:{password}@{host}:{port}/{database}"
+
+def setup_celery_connection(domain):
+    credentials = fetch_credentials(domain)
+    connection_string = serialize_connection(credentials)
+    connections.databases[domain] = dj_database_url.parse(connection_string)
+
```
Comment (johnmaia) [inline saleor/domain_utils.py]: Remove new line.

File: saleor/extensions/plugins/webhook/tasks.py [hunk 8]
```diff
@@ -8,15 +8,19 @@
 from ....webhook.models import Webhook
 from . import create_webhook_headers
```
Comment (johnmaia) [inline saleor/extensions/plugins/webhook/tasks.py]: Remove new line.

File: saleor/extensions/plugins/webhook/tasks.py [hunk 9]
```diff
@@ -39,11 +43,15 @@ def trigger_webhooks_for_event(event_type, data):
...
 @app.task(
+    base=DomainTask,
     autoretry_for=(RequestException,),
     retry_backoff=60,
     retry_kwargs={"max_retries": 15},
 )
-def send_webhook_request(webhook_id, target_url, secret, event_type, data):
+def send_webhook_request(webhook_id, target_url, secret, event_type, data, **kwargs):
+    domain = kwargs.get('domain')
+    setup_celery_connection(domain)
```
Comment (johnmaia) [inline saleor/extensions/plugins/webhook/tasks.py:57]: Add new line.

File: saleor/extensions/plugins/webhook/tasks.py [hunk 10]
```diff
@@ -8,15 +8,19 @@
...
 
 
-@app.task
-def trigger_webhooks_for_event(event_type, data):
-    permissions = {}
-    required_permission = WebhookEventType.PERMISSIONS[event_type].value
+@app.task(base=DomainTask)
+def trigger_webhooks_for_event(event_type, data, **kwargs):
+    domain = kwargs.get('domain')
+    setup_celery_connection(domain)
```
Comment (johnmaia) [inline saleor/extensions/plugins/webhook/tasks.py:22]: Add new line.

File: saleor/graphql/account/mutations/base.py [hunk 11]
```diff
@@ -26,6 +25,8 @@
 )
 from ...core.types.common import AccountError
```
Comment (johnmaia) [inline saleor/graphql/account/mutations/base.py]: Remove new line.

File: saleor/graphql/checkout/mutations.py [hunk 12]
```diff
@@ -50,6 +49,8 @@
 from ..shipping.types import ShippingMethod
 from .types import Checkout, CheckoutLine
```
Comment (johnmaia) [inline saleor/graphql/checkout/mutations.py]: Remove new line.

File: saleor/menu/utils.py [hunk 13]
```diff
@@ -1,7 +1,6 @@
-from django.db import transaction
-
 from ..menu.models import Menu
```
Comment (johnmaia) [inline saleor/menu/utils.py]: Remove new line.

File: saleor/migrate/tasks.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
+import dj_database_url
+
+from faker import Factory
+
```
Comment (johnmaia) [inline saleor/migrate/tasks.py]: Remove new line.

File: saleor/migrate/tasks.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+from celery import Task, signals
+from ..celeryconf import app
```
Comment (johnmaia) [inline saleor/migrate/tasks.py]: Let's organize the imports like the following: 
```py
import foobar
import foobiz

from fred import 
from waldo import

from ...quux import
from ...qux import
from ..bar import
from ..biz import
from .corge import
from .garply import
```

That seems to be how the team from saleor organizes the imports.

File: saleor/migrate/tasks.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+from django.db import connections
+from django.conf import settings
+from django.utils.text import slugify
+from django.core.management import call_command
+
+from ..shipping.models import ShippingZone
+from ..account.models import Address, User
+from ..warehouse.models import Stock, Warehouse
+
```
Comment (johnmaia) [inline saleor/migrate/tasks.py]: Remove new line.

File: saleor/migrate/tasks.py (hunk 15 above)
Comment (johnmaia) [reply saleor/migrate/tasks.py]: And sort the three groups separately.

File: saleor/migrate/views.py [hunk 17]
```diff
@@ -0,0 +1,20 @@
+import json
+from django.http import HttpResponse, JsonResponse
+
```
Comment (johnmaia) [inline saleor/migrate/views.py]: Remove new line.

File: saleor/migrate/views.py [hunk 18]
```diff
@@ -0,0 +1,20 @@
+import json
+from django.http import HttpResponse, JsonResponse
```
Comment (johnmaia) [inline saleor/migrate/views.py:2]: Add new line.

File: saleor/order/actions.py [hunk 19]
```diff
@@ -19,6 +17,8 @@
     update_order_status,
 )
 
+from ..domain_utils import transaction_domain_atomic
```
Comment (johnmaia) [inline saleor/order/actions.py]: Move this import  to be declared after `from ..core` and before `from ..extensions.manager`

File: saleor/order/utils.py [hunk 20]
```diff
@@ -18,6 +17,7 @@
 from ..warehouse.management import allocate_stock, deallocate_stock, increase_stock
 from . import events
 
+from ..domain_utils import transaction_domain_atomic
```
Comment (johnmaia) [inline saleor/order/utils.py]: Same as above.

File: saleor/payment/utils.py [hunk 21]
```diff
@@ -15,6 +14,8 @@
 from .interface import AddressData, GatewayResponse, PaymentData
 from .models import Payment, Transaction
 
+from ..domain_utils import transaction_domain_atomic
```
Comment (johnmaia) [inline saleor/payment/utils.py]: Same as above.

File: saleor/product/thumbnails.py [hunk 22]
```diff
@@ -2,16 +2,24 @@
 from ..core.utils import create_thumbnails
 from .models import Category, Collection, ProductImage
 
+from ..domain_task import DomainTask
```
Comment (johnmaia) [inline saleor/product/thumbnails.py:3]: Same as above.

File: saleor/product/thumbnails.py [hunk 23]
```diff
@@ -2,16 +2,24 @@
...
+from ..domain_utils import setup_celery_connection
```
Comment (johnmaia) [inline saleor/product/thumbnails.py:4]: Same as above.

File: saleor/product/utils/__init__.py [hunk 24]
```diff
@@ -2,11 +2,12 @@
 from urllib.parse import urlencode
 
 from django.conf import settings
-from django.db import transaction
 
 from ...core.taxes import TaxedMoney, zero_taxed_money
 from ..tasks import update_products_minimal_variant_prices_task
 
+from ...domain_utils import transaction_domain_atomic
```
Comment (johnmaia) [inline saleor/product/utils/__init__.py:7]: Same as above.

File: saleor/settings.py [hunk 25]
```diff
@@ -10,6 +10,7 @@
 from django_prices.utils.formatting import get_currency_fraction
 from sentry_sdk.integrations.django import DjangoIntegration
```
Comment (johnmaia) [inline saleor/settings.py:11]: Remove this new line.

File: saleor/settings.py [hunk 26]
```diff
@@ -558,3 +554,7 @@ def get_host():
         "Make sure you've added storefront address to ALLOWED_CLIENT_HOSTS "
         "if ENABLE_ACCOUNT_CONFIRMATION_BY_EMAIL is enabled."
     )
+
+DOMAIN = {}
+
```
Comment (johnmaia) [inline saleor/settings.py]: Remove new line.

File: saleor/warehouse/management.py [hunk 27]
```diff
@@ -1,14 +1,14 @@
 from typing import TYPE_CHECKING
 
-from django.db import transaction
-
 from .models import Stock
 
+from ..domain_utils import transaction_domain_atomic
```
Comment (johnmaia) [inline saleor/warehouse/management.py:2]: Move this to be before `from .models` import.

File: saleor/wishlist/models.py [hunk 28]
```diff
@@ -1,10 +1,11 @@
 import uuid
 
-from django.db import models, transaction
+from django.db import models
 
 from ..account.models import User
 from ..product.models import Product, ProductVariant
```
Comment (johnmaia) [inline saleor/wishlist/models.py]: Remove new line.

**Review by johnmaia — CHANGES_REQUESTED** [review]
@dsbrgg please take a look at my comments.

File: saleor/extensions/plugins/webhook/tasks.py (hunk 8 above)
Comment (dsbrgg) [reply saleor/extensions/plugins/webhook/tasks.py]: This was not altered. Should the line be removed anyway?

File: saleor/graphql/account/mutations/base.py (hunk 11 above)
Comment (dsbrgg) [reply saleor/graphql/account/mutations/base.py]: This was not altered. Should the line be removed anyway?

File: saleor/graphql/checkout/mutations.py (hunk 12 above)
Comment (dsbrgg) [reply saleor/graphql/checkout/mutations.py]: This was not altered. Should the line be removed anyway?

File: saleor/menu/utils.py (hunk 13 above)
Comment (dsbrgg) [reply saleor/menu/utils.py]: This was not altered. Should the line be removed anyway?

File: saleor/settings.py (hunk 25 above)
Comment (dsbrgg) [reply saleor/settings.py:11]: This was not altered. Should the line be removed anyway?

Comment (johnmaia) [reply saleor/settings.py:11]: It wasn't altered because you didn't add the import right after the last one.

Comment (dsbrgg) [reply saleor/settings.py:11]: Whoops 😅

**Review by johnmaia — APPROVED** [review]

**Pull request merged**